extensions/amp-story/1.0/amp-story-parallax-service.js

```diff
@@ -0,0 +1,427 @@
+        setStyles(this.element, {
+          perspectiveOrigin: `${Math.round(
+            x * AMPLIFICATION_FACTOR
+          )}px ${Math.round(y * AMPLIFICATION_FACTOR)}px`,
+        });
+      });
+    });
+  }
+
+  /**
+   * Resets any styles applied during the updates
+   * @return {!Promise}
+   */
+  resetStyles() {
+    const layers = this.getLayers();
+
+    return this.vsync_
+      .mutatePromise(() => {
+        resetStyles(this.element, ['perspectiveOrigin', 'perspective']);
+
+        layers.forEach(layer => {
+          resetStyles(layer, ['contain', 'overflow']);
+        });
+      })
+      .then(() => {
+        // Reset the transitions after all other styles are reset to avoid
+        // animating the reset
+        return this.vsync_.mutatePromise(() => {
+          resetStyles(this.element, ['transition']);
+        });
+      })
+      .then(() => {
+        this.initialized_ = false;
+      });
+  }
+
+  /**
+   * Initializes the page's layers by giving them the appropriate translation and scaling
+   * @return {!Promise}
+   */
+  setInitialStyles() {
+    if (!this.shouldUpdate() || this.initialized_) {
+      return Promise.resolve();
+    }
+
+    return this.vsync_.mutatePromise(() => {
+      const layers = this.getLayers();
+
+      setStyles(this.element, {
+        perspective: PERSPECTIVE * layers.length + 'px',
+      });
+
+      setImportantStyles(this.element, {
+        transition: 'opacity 350ms cubic-bezier(0.0,0.0,0.2,1)',
+      });
+
+      const originLayerIndex = findIndex(layers, layer =>
+        layer.hasAttribute(ORIGIN_LAYER_ATTR)
+      );
+
+      const layerZIndexOffset =
+        originLayerIndex == -1
+          ? Math.round(layers.length / 2)
+          : originLayerIndex;
+
+      // Loop through the layers in the page and assign a z-index following
+      // DOM order (manual override will be added in the future)
+      layers.forEach((layer, order) => {
+        const translation = `translateZ(${(order - layerZIndexOffset) *
+          this.layerSpacing_ *
+          (PERSPECTIVE / AMPLIFICATION_FACTOR)}px)`;
+        const scale = `scale(${mapRange(
+          order,
+          0,
+          layers.length - 1,
+          this.farthestScale_,
+          this.nearestScale_
+        )})`;
+        setImportantStyles(layer, {
+          contain: 'none',
+          overflow: 'visible',
+          transform: `${translation} ${scale}`,
+        });
+      });
+
+      this.initialized_ = true;
+    });
+  }
+
+  /**
+   * True if the element is in the viewport.
+   * @return {boolean}
+   */
+  shouldUpdate() {
+    return this.element.hasAttribute('active');
+  }
+
+  /**
+   * Discovers and returns all layers inside a page
+   * @return {!Array<!Element>}
+   */
+  getLayers() {
+    return toArray(this.element.querySelectorAll(`amp-story-grid-layer`))
+      .map(layer => dev().assertElement(layer))
+      .filter(layer => !layer.hasAttribute(NO_PARALLAX_FX_ATTR));
```

Comment:
Nit: `toArray(this.element.querySelectorAll('amp-story-grid-layer:not([${NO_PARALLAX_FX_ATTR}])'))`
(you might have to use `escapeCssSelectorIdent`)

Reply:
Wait, I thought `no-parallax-fx` was an attribute on the page, not the layer?  Does this need to be filtered at all?  Wouldn't we just not process pages that have the `NO_PARALLAX_FX_ATTR` on them?

Reply:
@newmuis Hmm not sure I follow. This is the opt out attribute which works on the page level as well as the layer level (you can choose to not to animate specific pages or specific layers).

Reply:
Ah, ok, I wasn't aware you could opt layers out.  I'm not sure if this might look broken?  e.g. if you have 4 layers (from Layer A all the way at the back to Layer D in front), then if you put `no-parallax-fx` on Layer B, I would expect this would look very odd.

Reply:
Yes, the behavior is still TBD :/ We'll need UX to figure out what developers would want it to mean (have the z offset but no moving vs disabling all styles)

Reply:
I think it's probably simpler to just not allow opting layers out?  What's the use case there?

Reply:
@hongwei1990 for comments on this

Reply:
As to not block this I've added a TODO for it. There are more PRs coming for this experiment anyway.